### ISSUE TITLE
[BUG] DeepSpeed-Inference, AutoTP, single-node, multi-gpu: random output from Llama-3.1-8B-Instruct

### ISSUE THREAD

**Issue by songdezhao** [deepspeedai/DeepSpeed#7278]
**Describe the bug**
I was using DeepSpeed-Inference to do inference with Llama-3.1-8B-Instruct.

I did this on a single node with 80GB*8 GPUs, using AutoTP and Accelerate's split_between_processes.

The code works and I can see the model was correctly sharded across the 8 GPUs; however, the output seems to be some random text.

**To Reproduce**
1. Simple inference script to reproduce
Here is the code snippet. Here I have two inputs but 8 GPUs, so the inputs will be padded to 8 total elements.

What I found interesting is that if I replace "inputs" with ["What is huggingface", "What is huggingface",], then it works. Basically, it works if all elements in the "inputs" list are the same.

```
import os
from datetime import timedelta

import deepspeed
import torch
from accelerate import PartialState
from deepspeed.runtime.utils import see_memory_usage
from transformers import AutoModelForCausalLM, AutoTokenizer

deepspeed.init_distributed(timeout=timedelta(seconds=200000))
state = PartialState()

# download llama-3.1-8b-instruct to local directory: "model_to_evaluate"
model_path = "./model_to_evaluate"

# load tokenizer
tokenizer = AutoTokenizer.from_pretrained(model_path, add_bos_token=True, add_eos_token=False)
state.wait_for_everyone()
print(f"process: {state.process_index}, tokenizer loaded")

# load model to cpu
kwargs = {"torch_dtype": torch.float16, "attn_implementation": "flash_attention_2"}
model = AutoModelForCausalLM.from_pretrained(model_path, **kwargs)
state.wait_for_everyone()
see_memory_usage(f"process: {state.process_index}, after loading model to cpu", force=True)

# initialize deepspeed-inference
ds_inference_config = {
    "dtype": kwargs["torch_dtype"],
    "replace_with_kernel_inject": False,
    "tensor_parallel": {
        "tp_size": int(os.getenv("WORLD_SIZE", "1"))
    }
}
ds_engine = deepspeed.init_inference(model, config=ds_inference_config)
model = ds_engine.module
model.eval()
state.wait_for_everyone()
see_memory_usage(f"process: {state.process_index}, after deepspeed-inference initialization", force=True)
print(f"process: {state.process_index}, "
      f"ds config: {ds_inference_config}, "
      f"model device: {next(model.parameters()).device}")

inputs = [
    "What is DeepSpeed",
    "What is huggingface",
    # "What is huggingface",
]
inputs_tokenized = [
    tokenizer.apply_chat_template([{"role": "user", "content": text}], add_generation_prompt=True, tokenize=False)
    for text in inputs
]
print(f"process: {state.process_index}, all inputs before split:\n{inputs_tokenized}")

with state.split_between_processes(inputs=inputs_tokenized, apply_padding=True) as inputs_for_process:
    print(f"process: {state.process_index}, num_samples: {len(inputs_for_process)}")

    model_inputs = tokenizer(inputs_for_process, return_tensors="pt", add_special_tokens=False)
    print(f"process: {state.process_index}, "
          f"first tokenized sample: {tokenizer.decode(model_inputs['input_ids'][0], skip_special_tokens=False)}")

    model_inputs = model_inputs.to(state.device)

    generated_ids = model.generate(**model_inputs,
                                   pad_token_id=tokenizer.pad_token_id,
                                   eos_token_id=tokenizer.eos_token_id,
                                   max_new_tokens=100,
                                   synced_gpus=True,
                                   use_cache=True,
                                   do_sample=False)

    prompt_length = model_inputs['input_ids'].shape[1]
    batch_responses = tokenizer.batch_decode(generated_ids[:, prompt_length:], skip_special_tokens=True)
    print(f"process: {state.process_index}/{state.device}, batch responses: {batch_responses}")
state.wait_for_everyone()
print(f"process: {state.process_index}, finished")
```

2. What packages are required and their versions
```
transformers==4.51.3
accelerate==1.6.0
deepspeed==0.16.7
flash-attn==2.7.3
torch==2.6.0+cu126
```

3. How to run the script
Step 1: Download Llama-3.1-8B-Instruct from huggingface into this local directory: "model_to_evaluate".
Step 2: Put the above code to "test_multi_gpu.py".
Step 3: Run this: accelerate launch test_multi_gpu.py

The output text will be like this on all processes (the order of the processes may differ):
process: 0/cuda:0, batch responses: [' (1, 2, 3, 4, 5)']
process: 1/cuda:1, batch responses: [' (1, 2, 3, 4, 5)']
process: 2/cuda:2, batch responses: [' (1, 2, 3, 4, 5)']
process: 3/cuda:3, batch responses: [' (1, 2, 3, 4, 5)']
process: 4/cuda:4, batch responses: [' (1, 2, 3, 4, 5)']
process: 5/cuda:5, batch responses: [' (1, 2, 3, 4, 5)']
process: 6/cuda:6, batch responses: [' (1, 2, 3, 4, 5)']
process: 7/cuda:7, batch responses: [' (1, 2, 3, 4, 5)']

I am wondering whether I did anything wrong here?

4. ...
**Expected behavior**
Expecting the model to provide a good response to the "inputs" in the above code.

**ds_report output**
Please run `ds_report` to give us details about your setup.

**Screenshots**

**System info (please complete the following information):**
 - OS: [e.g. Ubuntu 18.04]
 - GPU count and types [e.g. two machines with x8 A100s each]
 - (if applicable) what [DeepSpeed-MII](https://github.com/deepspeedai/deepspeed-mii) version are you using
 - (if applicable) Hugging Face Transformers/Accelerate/etc. versions
 - Python version
 - Any other relevant info about your setup

**Docker context**
Are you using a specific docker image that you can share?

**Additional context**

**Comment by delock** [comment]
Does `inputs_tokenized` intended to be used?  I see it defined but not used in the sample code.

**Comment by delock** [comment]
@songdezhao I only have a CPU system aside so I trim down your test code to run on the CPU system, result seems normal under autotp

Output
```
...
['What is DeepSpeed?\nDeepSpeed is an open-source, high-performance deep learning library developed by Me
ta AI. It is designed to accelerate the training of large-scale deep learning models, such as transformer
-based models, on a variety of hardware platforms, including CPUs, GPUs, and TPUs.\nDeepSpeed provides a 
range of features and tools to optimize the training process, including:\n  1. Model parallelism: DeepSpe
ed allows you to split large models across multiple GPUs, reducing memory requirements and increasing tra
ining speed.\n ']                                                                                        
['What is DeepSpeed?\nDeepSpeed is an open-source, high-performance deep learning library developed by Me
ta AI. It is designed to accelerate the training of large-scale deep learning models, such as transformer
-based models, on a variety of hardware platforms, including CPUs, GPUs, and TPUs.\nDeepSpeed provides a 
range of features and tools to optimize the training process, including:\n  1. Model parallelism: DeepSpe
ed allows you to split large models across multiple GPUs, reducing memory requirements and increasing tra
ining speed.\n ']                                                                                        
...
```
Command to launch the code, I use deepspeed launcher instead of `accelerate` launcher
```
deepspeed --num_gpus 8 --bind_cores_to_rank test_multi_gpu.py    
```

Trimmed test, I set batch to 1 because for some reason I cannot get batch to 2, maybe something with padding.
```
import os
from datetime import timedelta

import deepspeed
import torch
from deepspeed.runtime.utils import see_memory_usage
from transformers import AutoModelForCausalLM, AutoTokenizer

# download llama-3.1-8b-instruct to local directory: "model_to_evaluate"
model_path = "./model_to_evaluate"

# load tokenizer
tokenizer = AutoTokenizer.from_pretrained(model_path, trust_remote_code=True)

# load model to cpu
kwargs = {"torch_dtype": torch.float16, "attn_implementation": "flash_attention_2"}
model = AutoModelForCausalLM.from_pretrained(model_path, trust_remote_code=True)

# initialize deepspeed-inference
ds_inference_config = {
    "dtype": kwargs["torch_dtype"],
    "replace_with_kernel_inject": False,
    "tensor_parallel": {
        "tp_size": int(os.getenv("WORLD_SIZE", "1"))
    }
}
ds_engine = deepspeed.init_inference(model, config=ds_inference_config)
model = ds_engine.module
model.eval()

inputs = [
    "What is DeepSpeed",
    #"What is huggingface",
    # "What is huggingface",
]

model_inputs = tokenizer(inputs, return_tensors="pt")

generated_ids = model.generate(**model_inputs,
                               pad_token_id=tokenizer.pad_token_id,
                               eos_token_id=tokenizer.eos_token_id,
                               max_new_tokens=100,
                               synced_gpus=True,
                               use_cache=True,
                               do_sample=False)

batch_responses = tokenizer.batch_decode(generated_ids, skip_special_tokens=True)
print (batch_responses)
```

**Comment by songdezhao** [comment]
Hello @delock , thanks for looking into this.

1) Yes, ```inputs_tokenized``` is being used in this line. Here, I use accelerate to split the inputs to multiple processes. There are only 2 input strings, so the ```apply_padding=True``` in this line will duplicate the second string to make it a total of 8 strings.
```
with state.split_between_processes(inputs=inputs_tokenized, apply_padding=True) as inputs_for_process:
```

2) I tried my original code with the deepspeed launcher on a single node with 80GB*8 GPUs and got the same random output as above.

3) Could you try my original code on a single node with multiple GPUs? Please also use the same inputs and also keep the line where I split the inputs between different processes.

Thanks again for looking into this.

**Comment by songdezhao** [comment]
@delock, thanks again for looking into this.

I am particularly interested in using deepspeed-inference to see it would be helpful for large models.

Compared to the other issues I raised, this specific issue is more important I think, since the output seems a bit unusual with autotp+split_between_processes.

**Comment by delock** [comment]
After apply split_between_process code the output looks like below.

If I understand correctly, `split_between_processes` would make each worker process different input, yet for autotp each rank should take exactly the same input.  So you have different input for different ranks while each rank expect to process the same input (the full batch), maybe that's where you get random input.  It also explains when you duplicate batch inputs you get normal output, in this each each rank get same input again.   The right way to do it should be not split the batch to process but just pad the batch.

```
['system\n\nCutting Knowledge Date: December 2023\nToday Date: 26 Jul 2024\n\nuser\n\nWhat is DeepSpeedassistant\n\nHugging Face is an open-source artificial intelligence (AI) model hub that provides a wide range of pre-trained models and a simple, user-fr
iendly interface for building, training, and deploying natural language processing (NLP) and computer vision models. The platform was founded in 2016 by Clement Delangue and Clément Soubiran.\n\nHugging Face offers a variety of features and tools, includin
g:\n\n1. **Transformers**: Hugging Face is particularly known for its Transformers library, which']
['system\n\nCutting Knowledge Date: December 2023\nToday Date: 26 Jul 2024\n\nuser\n\nWhat is huggingfaceassistant\n\nHugging Face is an open-source artificial intelligence (AI) model hub that provides a wide range of pre-trained models and a simple, user-
friendly interface for building, training, and deploying natural language processing (NLP) and computer vision models. The platform was founded in 2016 by Clement Delangue and Clément Soubiran.\n\nHugging Face offers a variety of features and tools, includ
ing:\n\n1. **Transformers**: Hugging Face is particularly known for its Transformers library, which']
['system\n\nCutting Knowledge Date: December 2023\nToday Date: 26 Jul 2024\n\nuser\n\nWhat is huggingfaceassistant\n\nHugging Face is an open-source artificial intelligence (AI) model hub that provides a wide range of pre-trained models and a simple, user-
friendly interface for building, training, and deploying natural language processing (NLP) and computer vision models. The platform was founded in 2016 by Clement Delangue and Clément Soubiran.\n\nHugging Face offers a variety of features and tools, includ
ing:\n\n1. **Transformers**: Hugging Face is particularly known for its Transformers library, which']
['system\n\nCutting Knowledge Date: December 2023\nToday Date: 26 Jul 2024\n\nuser\n\nWhat is huggingfaceassistant\n\nHugging Face is an open-source artificial intelligence (AI) model hub that provides a wide range of pre-trained models and a simple, user-
friendly interface for building, training, and deploying natural language processing (NLP) and computer vision models. The platform was founded in 2016 by Clement Delangue and Clément Soubiran.\n\nHugging Face offers a variety of features and tools, includ
ing:\n\n1. **Transformers**: Hugging Face is particularly known for its Transformers library, which']
['system\n\nCutting Knowledge Date: December 2023\nToday Date: 26 Jul 2024\n\nuser\n\nWhat is huggingfaceassistant\n\nHugging Face is an open-source artificial intelligence (AI) model hub that provides a wide range of pre-trained models and a simple, user-
friendly interface for building, training, and deploying natural language processing (NLP) and computer vision models. The platform was founded in 2016 by Clement Delangue and Clément Soubiran.\n\nHugging Face offers a variety of features and tools, includ
ing:\n\n1. **Transformers**: Hugging Face is particularly known for its Transformers library, which']
['system\n\nCutting Knowledge Date: December 2023\nToday Date: 26 Jul 2024\n\nuser\n\nWhat is huggingfaceassistant\n\nHugging Face is an open-source artificial intelligence (AI) model hub that provides a wide range of pre-trained models and a simple, user-
friendly interface for building, training, and deploying natural language processing (NLP) and computer vision models. The platform was founded in 2016 by Clement Delangue and Clément Soubiran.\n\nHugging Face offers a variety of features and tools, includ
ing:\n\n1. **Transformers**: Hugging Face is particularly known for its Transformers library, which']
['system\n\nCutting Knowledge Date: December 2023\nToday Date: 26 Jul 2024\n\nuser\n\nWhat is huggingfaceassistant\n\nHugging Face is an open-source artificial intelligence (AI) model hub that provides a wide range of pre-trained models and a simple, user-
friendly interface for building, training, and deploying natural language processing (NLP) and computer vision models. The platform was founded in 2016 by Clement Delangue and Clément Soubiran.\n\nHugging Face offers a variety of features and tools, includ
ing:\n\n1. **Transformers**: Hugging Face is particularly known for its Transformers library, which']
['system\n\nCutting Knowledge Date: December 2023\nToday Date: 26 Jul 2024\n\nuser\n\nWhat is huggingfaceassistant\n\nHugging Face is an open-source artificial intelligence (AI) model hub that provides a wide range of pre-trained models and a simple, user-
friendly interface for building, training, and deploying natural language processing (NLP) and computer vision models. The platform was founded in 2016 by Clement Delangue and Clément Soubiran.\n\nHugging Face offers a variety of features and tools, includ
ing:\n\n1. **Transformers**: Hugging Face is particularly known for its Transformers library, which']
```

**Comment by songdezhao** [comment]
Thanks a lot for checking into this. Closing this.
